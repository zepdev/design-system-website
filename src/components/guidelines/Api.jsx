/* eslint max-len: ["error", { "code": 140 }] */
import React from 'react'
import { makeStyles } from '@material-ui/core/styles'
import TreeView from '@material-ui/lab/TreeView'
import TreeItem from '@material-ui/lab/TreeItem'
import IconArrowChevronRight from '@zlab-de/zel-react-icons/IconArrowChevronRight'
import IconArrowChevronDown from '@zlab-de/zel-react-icons/IconArrowChevronDown'
import guidelines from '@zlab-de/zel-react/guidelines.json'
import clsx from 'clsx'

const useStyles = makeStyles(theme => ({
  root: {
    marginBottom: `${theme.space.xxl.rem}rem`,
  },
}))

function useType(value) {
  if (Array.isArray(value)) {
    return 'array'
  }
  if (value === null) {
    return 'null'
  }

  return typeof value
}

function useLabel(value, type) {
  switch (type) {
    case 'array':
      return `Array(${value.length})`
    case 'null':
      return 'null'
    case 'undefined':
      return 'undefined'
    case 'function':
      return `f ${value.name}()`
    case 'object':
      return 'Object'
    case 'string':
      return `"${value}"`
    case 'symbol':
      return `Symbol(${String(value)})`
    case 'bigint':
    case 'boolean':
    case 'number':
    default:
      return String(value)
  }
}

function useTokenType(type) {
  switch (type) {
    case 'object':
    case 'array':
      return 'comment'
    default:
      return type
  }
}

function ObjectEntryLabel({ objectKey, objectValue }) {
  const type = useType(objectValue)
  const label = useLabel(objectValue, type)
  const tokenType = useTokenType(type)

  return (
    <React.Fragment>
      {objectKey}: <span className={clsx('token', tokenType)}>{label}</span>
    </React.Fragment>
  )
}

function ObjectEntry(props) {
  const { nodeId, objectKey, objectValue } = props

  const keyPrefix = nodeId

  let children = null
  if (
    (objectValue !== null && typeof objectValue === 'object') ||
    typeof objectValue === 'function'
  ) {
    children =
      Object.keys(objectValue).length === 0
        ? undefined
        : Object.keys(objectValue).map(key => {
            return (
              <ObjectEntry
                key={key}
                nodeId={`${keyPrefix}.${key}`}
                objectKey={key}
                objectValue={objectValue[key]}
              />
            )
          })
  }

  return (
    <TreeItem
      nodeId={nodeId}
      label={
        <ObjectEntryLabel objectKey={objectKey} objectValue={objectValue} />
      }
    >
      {children}
    </TreeItem>
  )
}

function Api() {
  const classes = useStyles()

  return (
    <div className={classes.root}>
      <TreeView
        defaultCollapseIcon={<IconArrowChevronDown />}
        defaultExpandIcon={<IconArrowChevronRight />}
      >
        {Object.keys(guidelines).map(section => {
          return (
            <ObjectEntry
              key={section}
              nodeId={`${'ROOT'}.${section}`}
              objectKey={section}
              objectValue={guidelines[section]}
            />
          )
        })}
      </TreeView>
    </div>
  )
}

export default Api
