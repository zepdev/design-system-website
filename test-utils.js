import React from 'react'
import { render } from '@testing-library/react'
import { ThemeProvider } from '@material-ui/core/styles'
import theme from './src/theme'

const AllTheProviders = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>
}

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options })

// re-export everything
export * from '@testing-library/react'

// override render method
export { customRender as render }
