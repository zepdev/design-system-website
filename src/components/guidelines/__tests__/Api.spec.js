import React from "react"
import { render } from "../../../../test-utils"
import Api from "../Api"

describe("Api", () => {
  it("renders correctly", () => {
    const component = render(<Api />)
    expect(component).toMatchSnapshot()
  })
})
