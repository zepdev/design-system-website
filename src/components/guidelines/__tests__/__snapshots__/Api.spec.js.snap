// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Api renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="makeStyles-root-1"
      >
        <ul
          aria-multiselectable="false"
          class="MuiTreeView-root"
          id="mui-25663"
          role="tree"
          tabindex="0"
        >
          <li
            aria-expanded="false"
            class="MuiTreeItem-root"
            id="mui-25663-ROOT.breakpoint"
            role="treeitem"
            tabindex="-1"
          >
            <div
              class="MuiTreeItem-content"
            >
              <div
                class="MuiTreeItem-iconContainer"
              >
                <svg
                  aria-hidden="true"
                  data-testid="ZepIconArrowChevronRightIcon"
                  focusable="false"
                  height="24"
                  viewBox="0 0 24 24"
                  width="24"
                >
                  <path
                    d="M14.74 12c0 .21-.07.42-.2.59l-.08.09-4.14 4.14a.62.62 0 01-.95-.8l.07-.09L13.37 12 9.44 8.07a.63.63 0 01-.07-.8l.07-.09a.63.63 0 01.8-.07l.08.07 4.14 4.14c.18.18.28.42.28.68z"
                    fill="currentColor"
                    fill-rule="evenodd"
                    transform="scale(Infinity, Infinity)"
                  />
                </svg>
              </div>
              <div
                class="MuiTreeItem-label"
              >
                breakpoint
                : 
                <span
                  class="token comment"
                >
                  Object
                </span>
              </div>
            </div>
          </li>
          <li
            aria-expanded="false"
            class="MuiTreeItem-root"
            id="mui-25663-ROOT.grid"
            role="treeitem"
            tabindex="-1"
          >
            <div
              class="MuiTreeItem-content"
            >
              <div
                class="MuiTreeItem-iconContainer"
              >
                <svg
                  aria-hidden="true"
                  data-testid="ZepIconArrowChevronRightIcon"
                  focusable="false"
                  height="24"
                  viewBox="0 0 24 24"
                  width="24"
                >
                  <path
                    d="M14.74 12c0 .21-.07.42-.2.59l-.08.09-4.14 4.14a.62.62 0 01-.95-.8l.07-.09L13.37 12 9.44 8.07a.63.63 0 01-.07-.8l.07-.09a.63.63 0 01.8-.07l.08.07 4.14 4.14c.18.18.28.42.28.68z"
                    fill="currentColor"
                    fill-rule="evenodd"
                    transform="scale(Infinity, Infinity)"
                  />
                </svg>
              </div>
              <div
                class="MuiTreeItem-label"
              >
                grid
                : 
                <span
                  class="token comment"
                >
                  Object
                </span>
              </div>
            </div>
          </li>
          <li
            aria-expanded="false"
            class="MuiTreeItem-root"
            id="mui-25663-ROOT.typography"
            role="treeitem"
            tabindex="-1"
          >
            <div
              class="MuiTreeItem-content"
            >
              <div
                class="MuiTreeItem-iconContainer"
              >
                <svg
                  aria-hidden="true"
                  data-testid="ZepIconArrowChevronRightIcon"
                  focusable="false"
                  height="24"
                  viewBox="0 0 24 24"
                  width="24"
                >
                  <path
                    d="M14.74 12c0 .21-.07.42-.2.59l-.08.09-4.14 4.14a.62.62 0 01-.95-.8l.07-.09L13.37 12 9.44 8.07a.63.63 0 01-.07-.8l.07-.09a.63.63 0 01.8-.07l.08.07 4.14 4.14c.18.18.28.42.28.68z"
                    fill="currentColor"
                    fill-rule="evenodd"
                    transform="scale(Infinity, Infinity)"
                  />
                </svg>
              </div>
              <div
                class="MuiTreeItem-label"
              >
                typography
                : 
                <span
                  class="token comment"
                >
                  Object
                </span>
              </div>
            </div>
          </li>
          <li
            aria-expanded="false"
            class="MuiTreeItem-root"
            id="mui-25663-ROOT.theme"
            role="treeitem"
            tabindex="-1"
          >
            <div
              class="MuiTreeItem-content"
            >
              <div
                class="MuiTreeItem-iconContainer"
              >
                <svg
                  aria-hidden="true"
                  data-testid="ZepIconArrowChevronRightIcon"
                  focusable="false"
                  height="24"
                  viewBox="0 0 24 24"
                  width="24"
                >
                  <path
                    d="M14.74 12c0 .21-.07.42-.2.59l-.08.09-4.14 4.14a.62.62 0 01-.95-.8l.07-.09L13.37 12 9.44 8.07a.63.63 0 01-.07-.8l.07-.09a.63.63 0 01.8-.07l.08.07 4.14 4.14c.18.18.28.42.28.68z"
                    fill="currentColor"
                    fill-rule="evenodd"
                    transform="scale(Infinity, Infinity)"
                  />
                </svg>
              </div>
              <div
                class="MuiTreeItem-label"
              >
                theme
                : 
                <span
                  class="token comment"
                >
                  Object
                </span>
              </div>
            </div>
          </li>
          <li
            aria-expanded="false"
            class="MuiTreeItem-root"
            id="mui-25663-ROOT.logo"
            role="treeitem"
            tabindex="-1"
          >
            <div
              class="MuiTreeItem-content"
            >
              <div
                class="MuiTreeItem-iconContainer"
              >
                <svg
                  aria-hidden="true"
                  data-testid="ZepIconArrowChevronRightIcon"
                  focusable="false"
                  height="24"
                  viewBox="0 0 24 24"
                  width="24"
                >
                  <path
                    d="M14.74 12c0 .21-.07.42-.2.59l-.08.09-4.14 4.14a.62.62 0 01-.95-.8l.07-.09L13.37 12 9.44 8.07a.63.63 0 01-.07-.8l.07-.09a.63.63 0 01.8-.07l.08.07 4.14 4.14c.18.18.28.42.28.68z"
                    fill="currentColor"
                    fill-rule="evenodd"
                    transform="scale(Infinity, Infinity)"
                  />
                </svg>
              </div>
              <div
                class="MuiTreeItem-label"
              >
                logo
                : 
                <span
                  class="token comment"
                >
                  Object
                </span>
              </div>
            </div>
          </li>
          <li
            aria-expanded="false"
            class="MuiTreeItem-root"
            id="mui-25663-ROOT.color"
            role="treeitem"
            tabindex="-1"
          >
            <div
              class="MuiTreeItem-content"
            >
              <div
                class="MuiTreeItem-iconContainer"
              >
                <svg
                  aria-hidden="true"
                  data-testid="ZepIconArrowChevronRightIcon"
                  focusable="false"
                  height="24"
                  viewBox="0 0 24 24"
                  width="24"
                >
                  <path
                    d="M14.74 12c0 .21-.07.42-.2.59l-.08.09-4.14 4.14a.62.62 0 01-.95-.8l.07-.09L13.37 12 9.44 8.07a.63.63 0 01-.07-.8l.07-.09a.63.63 0 01.8-.07l.08.07 4.14 4.14c.18.18.28.42.28.68z"
                    fill="currentColor"
                    fill-rule="evenodd"
                    transform="scale(Infinity, Infinity)"
                  />
                </svg>
              </div>
              <div
                class="MuiTreeItem-label"
              >
                color
                : 
                <span
                  class="token comment"
                >
                  Object
                </span>
              </div>
            </div>
          </li>
          <li
            aria-expanded="false"
            class="MuiTreeItem-root"
            id="mui-25663-ROOT.spacing"
            role="treeitem"
            tabindex="-1"
          >
            <div
              class="MuiTreeItem-content"
            >
              <div
                class="MuiTreeItem-iconContainer"
              >
                <svg
                  aria-hidden="true"
                  data-testid="ZepIconArrowChevronRightIcon"
                  focusable="false"
                  height="24"
                  viewBox="0 0 24 24"
                  width="24"
                >
                  <path
                    d="M14.74 12c0 .21-.07.42-.2.59l-.08.09-4.14 4.14a.62.62 0 01-.95-.8l.07-.09L13.37 12 9.44 8.07a.63.63 0 01-.07-.8l.07-.09a.63.63 0 01.8-.07l.08.07 4.14 4.14c.18.18.28.42.28.68z"
                    fill="currentColor"
                    fill-rule="evenodd"
                    transform="scale(Infinity, Infinity)"
                  />
                </svg>
              </div>
              <div
                class="MuiTreeItem-label"
              >
                spacing
                : 
                <span
                  class="token comment"
                >
                  Object
                </span>
              </div>
            </div>
          </li>
        </ul>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="makeStyles-root-1"
    >
      <ul
        aria-multiselectable="false"
        class="MuiTreeView-root"
        id="mui-25663"
        role="tree"
        tabindex="0"
      >
        <li
          aria-expanded="false"
          class="MuiTreeItem-root"
          id="mui-25663-ROOT.breakpoint"
          role="treeitem"
          tabindex="-1"
        >
          <div
            class="MuiTreeItem-content"
          >
            <div
              class="MuiTreeItem-iconContainer"
            >
              <svg
                aria-hidden="true"
                data-testid="ZepIconArrowChevronRightIcon"
                focusable="false"
                height="24"
                viewBox="0 0 24 24"
                width="24"
              >
                <path
                  d="M14.74 12c0 .21-.07.42-.2.59l-.08.09-4.14 4.14a.62.62 0 01-.95-.8l.07-.09L13.37 12 9.44 8.07a.63.63 0 01-.07-.8l.07-.09a.63.63 0 01.8-.07l.08.07 4.14 4.14c.18.18.28.42.28.68z"
                  fill="currentColor"
                  fill-rule="evenodd"
                  transform="scale(Infinity, Infinity)"
                />
              </svg>
            </div>
            <div
              class="MuiTreeItem-label"
            >
              breakpoint
              : 
              <span
                class="token comment"
              >
                Object
              </span>
            </div>
          </div>
        </li>
        <li
          aria-expanded="false"
          class="MuiTreeItem-root"
          id="mui-25663-ROOT.grid"
          role="treeitem"
          tabindex="-1"
        >
          <div
            class="MuiTreeItem-content"
          >
            <div
              class="MuiTreeItem-iconContainer"
            >
              <svg
                aria-hidden="true"
                data-testid="ZepIconArrowChevronRightIcon"
                focusable="false"
                height="24"
                viewBox="0 0 24 24"
                width="24"
              >
                <path
                  d="M14.74 12c0 .21-.07.42-.2.59l-.08.09-4.14 4.14a.62.62 0 01-.95-.8l.07-.09L13.37 12 9.44 8.07a.63.63 0 01-.07-.8l.07-.09a.63.63 0 01.8-.07l.08.07 4.14 4.14c.18.18.28.42.28.68z"
                  fill="currentColor"
                  fill-rule="evenodd"
                  transform="scale(Infinity, Infinity)"
                />
              </svg>
            </div>
            <div
              class="MuiTreeItem-label"
            >
              grid
              : 
              <span
                class="token comment"
              >
                Object
              </span>
            </div>
          </div>
        </li>
        <li
          aria-expanded="false"
          class="MuiTreeItem-root"
          id="mui-25663-ROOT.typography"
          role="treeitem"
          tabindex="-1"
        >
          <div
            class="MuiTreeItem-content"
          >
            <div
              class="MuiTreeItem-iconContainer"
            >
              <svg
                aria-hidden="true"
                data-testid="ZepIconArrowChevronRightIcon"
                focusable="false"
                height="24"
                viewBox="0 0 24 24"
                width="24"
              >
                <path
                  d="M14.74 12c0 .21-.07.42-.2.59l-.08.09-4.14 4.14a.62.62 0 01-.95-.8l.07-.09L13.37 12 9.44 8.07a.63.63 0 01-.07-.8l.07-.09a.63.63 0 01.8-.07l.08.07 4.14 4.14c.18.18.28.42.28.68z"
                  fill="currentColor"
                  fill-rule="evenodd"
                  transform="scale(Infinity, Infinity)"
                />
              </svg>
            </div>
            <div
              class="MuiTreeItem-label"
            >
              typography
              : 
              <span
                class="token comment"
              >
                Object
              </span>
            </div>
          </div>
        </li>
        <li
          aria-expanded="false"
          class="MuiTreeItem-root"
          id="mui-25663-ROOT.theme"
          role="treeitem"
          tabindex="-1"
        >
          <div
            class="MuiTreeItem-content"
          >
            <div
              class="MuiTreeItem-iconContainer"
            >
              <svg
                aria-hidden="true"
                data-testid="ZepIconArrowChevronRightIcon"
                focusable="false"
                height="24"
                viewBox="0 0 24 24"
                width="24"
              >
                <path
                  d="M14.74 12c0 .21-.07.42-.2.59l-.08.09-4.14 4.14a.62.62 0 01-.95-.8l.07-.09L13.37 12 9.44 8.07a.63.63 0 01-.07-.8l.07-.09a.63.63 0 01.8-.07l.08.07 4.14 4.14c.18.18.28.42.28.68z"
                  fill="currentColor"
                  fill-rule="evenodd"
                  transform="scale(Infinity, Infinity)"
                />
              </svg>
            </div>
            <div
              class="MuiTreeItem-label"
            >
              theme
              : 
              <span
                class="token comment"
              >
                Object
              </span>
            </div>
          </div>
        </li>
        <li
          aria-expanded="false"
          class="MuiTreeItem-root"
          id="mui-25663-ROOT.logo"
          role="treeitem"
          tabindex="-1"
        >
          <div
            class="MuiTreeItem-content"
          >
            <div
              class="MuiTreeItem-iconContainer"
            >
              <svg
                aria-hidden="true"
                data-testid="ZepIconArrowChevronRightIcon"
                focusable="false"
                height="24"
                viewBox="0 0 24 24"
                width="24"
              >
                <path
                  d="M14.74 12c0 .21-.07.42-.2.59l-.08.09-4.14 4.14a.62.62 0 01-.95-.8l.07-.09L13.37 12 9.44 8.07a.63.63 0 01-.07-.8l.07-.09a.63.63 0 01.8-.07l.08.07 4.14 4.14c.18.18.28.42.28.68z"
                  fill="currentColor"
                  fill-rule="evenodd"
                  transform="scale(Infinity, Infinity)"
                />
              </svg>
            </div>
            <div
              class="MuiTreeItem-label"
            >
              logo
              : 
              <span
                class="token comment"
              >
                Object
              </span>
            </div>
          </div>
        </li>
        <li
          aria-expanded="false"
          class="MuiTreeItem-root"
          id="mui-25663-ROOT.color"
          role="treeitem"
          tabindex="-1"
        >
          <div
            class="MuiTreeItem-content"
          >
            <div
              class="MuiTreeItem-iconContainer"
            >
              <svg
                aria-hidden="true"
                data-testid="ZepIconArrowChevronRightIcon"
                focusable="false"
                height="24"
                viewBox="0 0 24 24"
                width="24"
              >
                <path
                  d="M14.74 12c0 .21-.07.42-.2.59l-.08.09-4.14 4.14a.62.62 0 01-.95-.8l.07-.09L13.37 12 9.44 8.07a.63.63 0 01-.07-.8l.07-.09a.63.63 0 01.8-.07l.08.07 4.14 4.14c.18.18.28.42.28.68z"
                  fill="currentColor"
                  fill-rule="evenodd"
                  transform="scale(Infinity, Infinity)"
                />
              </svg>
            </div>
            <div
              class="MuiTreeItem-label"
            >
              color
              : 
              <span
                class="token comment"
              >
                Object
              </span>
            </div>
          </div>
        </li>
        <li
          aria-expanded="false"
          class="MuiTreeItem-root"
          id="mui-25663-ROOT.spacing"
          role="treeitem"
          tabindex="-1"
        >
          <div
            class="MuiTreeItem-content"
          >
            <div
              class="MuiTreeItem-iconContainer"
            >
              <svg
                aria-hidden="true"
                data-testid="ZepIconArrowChevronRightIcon"
                focusable="false"
                height="24"
                viewBox="0 0 24 24"
                width="24"
              >
                <path
                  d="M14.74 12c0 .21-.07.42-.2.59l-.08.09-4.14 4.14a.62.62 0 01-.95-.8l.07-.09L13.37 12 9.44 8.07a.63.63 0 01-.07-.8l.07-.09a.63.63 0 01.8-.07l.08.07 4.14 4.14c.18.18.28.42.28.68z"
                  fill="currentColor"
                  fill-rule="evenodd"
                  transform="scale(Infinity, Infinity)"
                />
              </svg>
            </div>
            <div
              class="MuiTreeItem-label"
            >
              spacing
              : 
              <span
                class="token comment"
              >
                Object
              </span>
            </div>
          </div>
        </li>
      </ul>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
